scripts/nanopb/generate_protos.sh

```diff
@@ -24,41 +24,48 @@ readonly DIR="$( git rev-parse --show-toplevel )"
 
 # Current release of nanopb being used  to build the CCT protos
 readonly NANOPB_VERSION="0.3.9.8"
-readonly NANOPB_TEMPDIR="${DIR}/FirebaseSessions/nanopb_temp"
+readonly NANOPB_TEMPDIR="${DIR}/scripts/nanopb/nanopb_temp"
 
-readonly LIBRARY_DIR="${DIR}/FirebaseSessions/Sources/"
-readonly PROTO_DIR="${DIR}/FirebaseSessions/ProtoSupport/Protos/"
-readonly PROTOGEN_DIR="${DIR}/FirebaseSessions/Protogen/"
+readonly PROTO_DIR="$1"
+readonly PROTOGEN_DIR="$2"
+readonly INCLUDE_PREFIX="$3"
+
+echoColor() {
+  COLOR='\033[0;35m'
+  NC='\033[0m'
+  printf "${COLOR}$1${NC}\n"
+}
 
 rm -rf "${NANOPB_TEMPDIR}"
 
-echo "Downloading nanopb..."
+echoColor "Downloading nanopb..."
 git clone --branch "${NANOPB_VERSION}" https://github.com/nanopb/nanopb.git "${NANOPB_TEMPDIR}"
 
-echo "Building nanopb..."
+echoColor "Building nanopb..."
 pushd "${NANOPB_TEMPDIR}"
 ./tools/make_mac_package.sh
 GIT_DESCRIPTION=`git describe --always`-macosx-x86
 NANOPB_BIN_DIR="dist/${GIT_DESCRIPTION}"
 popd
 
-echo "Removing existing protos..."
+echoColor "Removing existing protos..."
 rm -rf "${PROTOGEN_DIR}/*"
 
-echo "Generating protos..."
-python "${DIR}/FirebaseSessions/ProtoSupport/proto_generator.py" \
+echoColor "Generating protos..."
+python "${DIR}/scripts/nanopb/proto_generator.py" \
   --nanopb \
   --protos_dir="${PROTO_DIR}" \
   --pythonpath="${NANOPB_TEMPDIR}/${NANOPB_BIN_DIR}/generator" \
   --output_dir="${PROTOGEN_DIR}" \
-  --include="${PROTO_DIR}"
+  --include="${PROTO_DIR}" \
+  --include_prefix="${INCLUDE_PREFIX}"
 
 rm -rf "${NANOPB_TEMPDIR}"
 
 RED='\033[0;31m'
 NC='\033[0m'
 echo ""
 echo ""
-echo -e "${RED}Important: Any new proto fields of type string, repeated, or bytes must be specified in the sessions.options file${NC}"
+echo -e "${RED}Important: Any new proto fields of type string, repeated, or bytes must be specified in the proto's .options file with type:FT_POINTER${NC}"
```

Comment:
Curious, could you explain why the `FT_POINTER` type is needed? It sounds like at one point that it wasn't. I've seen and used it before but didn't know why.

Reply:
This change is more of a clarification to the original intent of this log. I added this log around a week ago.

NanoPB has a few options for this, and it defaults to FT_CALLBACK. FT_CALLBACK requires you to implement an encode and decode function for each field, versus FT_POINTER just generates a field that's a `void *` that you fill in.

Eg., I changed `session_id` to FT_CALLBACK, and here's the diff:

```
 typedef struct _firebase_appquality_sessions_SessionInfo {
-    pb_bytes_array_t *session_id;
+    pb_callback_t session_id;
     pb_bytes_array_t *previous_session_id;
     pb_bytes_array_t *firebase_installation_id;
```

pb_callback_t looks like, which is extra work to implement:
```
struct pb_callback_s {
#ifdef PB_OLD_CALLBACK_STYLE
    /* Deprecated since nanopb-0.2.1 */
    union {
        bool (*decode)(pb_istream_t *stream, const pb_field_t *field, void *arg);
        bool (*encode)(pb_ostream_t *stream, const pb_field_t *field, const void *arg);
    } funcs;
#else
    /* New function signature, which allows modifying arg contents in callback. */
    union {
        bool (*decode)(pb_istream_t *stream, const pb_field_t *field, void **arg);
        bool (*encode)(pb_ostream_t *stream, const pb_field_t *field, void * const *arg);
    } funcs;
#endif    
    
    /* Free arg for use by callback */
    void *arg;
};
```

Reply:
thanks Sam